app/controllers/order_items_controller.rb

```diff
@@ -1,13 +1,79 @@
 class OrderItemsController < ApplicationController
-    def create
-    @product = Product.find(params[:product_id])
-    @order = current_order
+  def create
+    order = current_order
+    product = Product.find_by(id: params[:product_id])
 
-    @order_item = @order.order_items.new(quantity: params[:order_item][:quantity], 
-      product_id: @product.id)
-    @order_item.save
-    session[:order_id] = @order.id
+    previous_product = order.products.find_by(id: product.id)
 
-    redirect_to @product
+    if previous_product
+      previous_order_item = order.order_items.find_by(product_id: previous_product.id)
+      previous_quantity = previous_order_item.quantity
+      previous_order_item.quantity += (params[:order_item][:quantity]).to_i
+      if previous_order_item.quantity > previous_product.stock
+        previous_order_item.quantity = previous_product.stock
+        flash[:errors] = "There are only #{previous_product.stock} x #{previous_product.name} available for sale."
+      end
+      previous_order_item.save
+      flash[:success] = "You have added #{previous_order_item.quantity - previous_quantity} x #{product.name} to your cart."
+    else # new product
+      order_item = order.order_items.new(order_item_params)
+      order_item.product = product
+      if params[:order_item][:quantity].to_i > order_item.product.stock
+        flash[:errors] = "You can't buy that many. Stock available for this item: #{order_item.product.stock}."
+        redirect_to product and return
+      end
+      order_item.save
+      if order_item.save
+        flash[:success] = "You have added #{order_item.quantity} x #{product.name} to your cart."
+      else
+        flash[:errors] = "There was a problem with adding this item to your cart."
+        if order.order_items.count == 0
+          order.delete
```

Comment:
Interesting. What's the rationale behind destroying the Order at this point? Won't they just get a new Order instance when they attempt to shop again? What's wrong with keeping this order around?